pkg/captcha/captcha.go

```diff
@@ -0,0 +1,228 @@
+package captcha
+
+import (
+	"encoding/json"
+	"fmt"
+	"html/template"
+	"net/http"
+	"net/url"
+	"os"
+	"time"
+
+	"github.com/golang-jwt/jwt/v5"
+	"github.com/maxlerebourg/crowdsec-bouncer-traefik-plugin/pkg/cache"
+	"github.com/maxlerebourg/crowdsec-bouncer-traefik-plugin/pkg/logger"
+)
+
+type Client struct {
+	Provider            string
+	Template            string
+	TemplateFile        string
+	CompiledTemplate    *template.Template
+	FallbackRemediation string
+	Valid               bool
+	GracePeriod         int64
+	SiteKey             string
+	SecretKey           string
+	Cache               *cache.Client
+	Http                *http.Client
+}
+
+var (
+	ValidProviders = []string{"hcaptcha", "recaptcha", "turnstile"}
+	FrontendJS     = map[string]string{
+		"hcaptcha":  "https://hcaptcha.com/1/api.js",
+		"recaptcha": "https://www.google.com/recaptcha/api.js",
+		"turnstile": "https://challenges.cloudflare.com/turnstile/v0/api.js",
+	}
+	FrontendKey = map[string]string{
+		"hcaptcha":  "h-captcha",
+		"recaptcha": "g-recaptcha",
+		"turnstile": "cf-captcha",
+	}
+	ValidateEndpoints = map[string]string{
+		"hcaptcha":  "https://api.hcaptcha.com/siteverify",
+		"recaptcha": "https://www.google.com/recaptcha/api/siteverify",
+		"turnstile": "https://challenges.cloudflare.com/turnstile/v0/siteverify",
+	}
+	InvalidMethod        = "invalid:method"
+	InvalidConfiguration = "invalid:configuration"
+)
+
+func (c *Client) CompileTemplate() error {
+	var err error
+	if c.Template == "" && c.TemplateFile == "" {
+		return fmt.Errorf("no captcha template provided, captcha decisions will fallback to %s", c.FallbackRemediation)
+	}
+	if c.TemplateFile != "" {
+		b, err := os.ReadFile(c.TemplateFile)
+		if err != nil {
+			return err
+		}
+		c.Template = string(b)
+	}
+	c.CompiledTemplate, err = template.New("captcha").Parse(c.Template)
+	if err != nil {
+		return fmt.Errorf("error compiling captcha template: %s", err.Error())
+	}
+	c.Valid = true
+	return err
+}
+
+func (c *Client) Debug(message string) {
+	logger.Debug(fmt.Sprintf("captchaClient: %s", message))
+}
+
+func (c *Client) New(siteKey, secretKey, provider, template, templateFile, fallback string, gracePeriod int64) error {
+	c.SiteKey = siteKey
+	c.SecretKey = secretKey
+	c.Provider = provider
+	c.Template = template
+	c.TemplateFile = templateFile
+	c.FallbackRemediation = fallback
+	c.GracePeriod = gracePeriod
+	c.Http = &http.Client{
+		Transport: &http.Transport{
+			MaxIdleConns:    10,
+			IdleConnTimeout: 30 * time.Second,
+		},
+		Timeout: 10 * time.Second,
+	}
+	return c.CompileTemplate()
+}
+
+func (c *Client) ServeHTTP(rw http.ResponseWriter, r *http.Request, remoteIP string) {
+	if !c.Valid && c.FallbackRemediation == "ban" {
```

Comment:
Might move this to the `handleRemediation` function within `bouncer.go` to override the incoming remediation string. Makes sense to hanndle it there instead of within the captcha stuff itself, as it will invoke the cookie for no reason.